1-Draft/RFCNNN-Null-conditional-operators.md

```diff
@@ -0,0 +1,140 @@
+${x?}?='new value'
+```
+
+### Null coalescing operator - `??`
+
+If the left hand side is null then return the right hand side, else return the right hand side.
+
+```powershell
+$x = $null
+$x ?? 100
+100
+```
+
+```powershell
+$x = 'some value'
+$x ?? 100
+some value
+```
+
+```powershell
+$x? = 'some value'
+${x?}??100
+some value
+```
+
+### Null conditional member access operators - `?.` and `?[]`
+
+Null member access operators can used on scalar types and array types.
+Return the value of the accessed member if the variable is not null.
+If the value of the variable is null, then return null.
+
+```powershell
+$x = $null
+${x}?.property
+${x?}?.property
+
+${x}?[0]
+${x?}?[0]
+```
+
+## Specification
+
+The RFC proposes four new operators, the null assignment operator `?=`, the null coalescing operator `??`, and the two null conditional member access `?.` and `?[]`.
+
+### Null assignment operator - `?=`
+
+The `?=` operator check the left hand side to be null or undefined and assigns the value of right hand side if null, else the value of left hand side is not changed.
+
+```powershell
+$todaysDate ?= (Get-Date)
+```
+
+The value of `Get-Date` is assigned to `$todaysDate` if it is null or undefined, if not null, the value is not changed.
+
+### Null coalescing operator - `??`
+
+The `??` operator checks the left hand side to be null or undefined and returns the value of right hand side if null, else return the value of left hand side.
+
+```powershell
+(Get-Module Pester -ListAvailable) ?? (Install-Module Pester)
+```
+
+If the module `Pester` is not installed, then install the module.
+
+### Null conditional member access operators - `?.` and `?[]`
+
+The `?.` and `?[]` operators access the specified member if the value of the variable is not null.
+If it is null, then null is returned.
+
+```powershell
+${x}?.propname
+${x}?[0]
+```
+
+The property `propname` is accessed and it's value is returned only if `$x` is not null.
+Similarly, the indexer is used only if `$x` is not null.
+If `$x` is null, then null is returned.
+
+### Additional considerations
+
+The `?=` and `??` operators are binary operators and check for the value of the left hand side to be null or an undefined variable.
+They can be used with a space between the variable name and the to operator.
+If the left hand side is null, then `?=` assigns the right hand side to the variable while, the `??` operator returns the right hand side.
+When they are used without a space between variable name and the operator, a `{}` must be used around the variable name.
+
+The `?.` and `?[]` operators are member access operators and do not allow a space in between the variable name and the operator.
+
+PowerShell allows `?` as part of the variable name.
+Hence disambiguation is required when the operators are used without a space between the variable name and the operator.
+To disambiguate, the variables must use `{}` around the variable name like: `${x?}?.propertyName` or `${y}?[0]`.
```

Comment:
I think it's worth mentioning that asking C# to change their identifier token pattern wouldn't be a statistical question, it would be a hard no.

Because PowerShell is designed to be so tolerant, it makes it harder for us to introduce new syntaxes like this, so it's worth some discussion. But I think we should be mindful that PowerShell is a programming language, and generally programming languages proceed with syntax being concretely backward compatible, or if they are really forced to break it they break it in an unsubtle way.

Reply:
I agree, Rob.

However, if this is a representative sample of the community's scripts, and the percentage of people using this syntax is _that_ low, I think it prudent to jump ship on `?` being a variable character, as it were.

C# also already does not allow `?` in an identifier if I recall correctly, and if folks really want to use it, we have the pre-existing catch-all of `${}` being usable to call variables whatever you want to call them.

I really don't think we're losing anything at all here.

Reply:
> I think it's worth mentioning that asking C# to change their identifier token pattern wouldn't be a statistical question, it would be a hard no.

Is that a quote? Even if it is, separate dialects is another common hard no from them, yet we're getting nullable reference types. Nothing is a hard no if there is a compelling enough reason.  If they had initially made the mistake of allowing `?` in variable names, they'd have a potentially compelling reason to change their identifier token pattern.

Honestly this breaking change would be very easy to fix in an affected script and would affect very few folks.  Contrast to that is the whole `Get-Content -AsByteStream` vs `Get-Content -Encoding Byte`.  That breaking change effected a much broader amount of folks (we even ran into it in PSES) and is comparatively way more involved to fix.  I get that that change doesn't necessarily mean that anything of lesser impact will be accepted, but imo this has a much more compelling benefit behind it (I'm still not really sure what was gained from the `byte` change tbh).

Reply:
To add to the conversation, going back and reviewing documentation, it's not clear anywhere that you can use non-alphanumeric + underscore variable names.  It's a fluke that it was allowed in the first place.

Reply:
The member access operators have been moved to additional consideration. This RFC now focuses on `??` and `??=`.

Reply:
From the @PowerShell/powershell-committee point of view, we are not accepting changes that break existing users' right to have `?` (or whitespace) in their variable names.  We certainly wouldn't recommend it, but we would not support such a breaking change.

Personally, I prefer alignment with C# syntax where possible for these advanced operators as I expect most usage would be from C# devs writing PS scripts.  However, in the case of conditional member access, I would be open to alternate syntax if we can preserve users' right to variable naming and not require braces.

Reply:
I think we've rather shot ourselves in the foot there. There isn't really a better syntax available, since literally all other null-coalescing operators currently existing or proposed use sigils containing `?`

So really, we either have a kind of pointless break from consistency and have a totally different sigil (for which there are precisely zero current suggestions, and which also wouldn't help c# programmers whatsoever), or we make a minor breaking change that is liable to affect very few users and remove `?` as a variable character.

Or we just accept that we will never safely be able to invoke members without awkward try/catch syntax.

I don't see a better solution than changing the requirements for variable names, unfortunately. Every other solution presented seems pretty lacking in my opinion. If anyone has any suggestions they're more than welcome, but I don't think we'll see a better solution without totally reworking all forms of null-coalescing and/or ternary operators.

Given our stated objective is to duplicate c#'s syntax for these operators, I don't think we currently have a better option than also more closely aligning what we can use for an identifier with c#'s standards.

Reply:
Maybe it's time to talk more seriously about [optional feature support](https://github.com/PowerShell/PowerShell-RFC/pull/220) then. No point in penalizing the entire community for a few use cases. Most of us would just turn on the feature by default though.

Either that, or maybe we should start putting functionality under .ps2/.psm2 extensions so that we have a direction to go in as we start cleaning up some of our messes.

Reply:
@vexx32 perpetual legacy support for an infinitesimally population of people using an undocumented feature.. *the PowerShell way* unfortunately.

Reply:
Spock says, “Logic clearly dictates that the needs of the many outweigh the needs of the few.”

Captain Kirk answers, “Or the one.”

Reply:
> ... perpetual legacy support for an infinitesimal population of people using an undocumented feature.. the PowerShell way unfortunately.

Using ? in a variable name has been explicitly documented since at least 2011. The "PowerShell Language Specification 2.0", section 2.3.2, page 6:

>variable-characters:
> variable-character
> variable-characters   variable-character
>
> variable-character:
> A Unicode character of classes Lu, Ll, Lt, Lm, Lo, or Nd
> _   (The underscore character U+005F)
> ?
> 

It's also in the "PowerShell Language Specification 3.0", section 2.3.2, on page 7.

Reply:
Sure, but it seems pretty clear it's a very rarely _used_ feature.

I guess at the end of the day, it's a question about what we care more about. Do we care more about backwards-compatibility in what really is essentially an edge case that is rarely used, or do we care more about easing the transition for C# programmers and having a reasonably sensible syntax for null-conditional member access that does deviate wildly from our standard property-access syntax?

Unless you'd prefer to have C#-oriented folks hopelessly confused and opt for `$var.?method()` instead, which is _just similar enough_ to look about right, but _just different enough_ to be hopelessly confusing.

Reply:
@essentialexch I'll admit I didn't look at the language spec since that's not where people go for help.  They look at the `about_` topics and from that point of view, it was not documented there.

Reply:
Me personally? I've primarily stayed out of this discussion after the crap-show that was the ternary discussion.

While I see the utility of these operators and appreciate the (potential?) performance gain, I do not approve of the obtuseness of the operators or the fact (what I see as a fact) that these things are turning the language into an unreadable mess, just like perl.

An admin scripter is going to have no idea what these things do. He/she isn't going to understand when they work, why they work, or what they mean. There is just going to be lots more garbage copy-and-pasted code all over the web.

I don't see this about tech. debt - I see it about PowerShell being a scripting language. What's next? Templates?

I'd prefer something far more verbose and obvious in meaning.

Reply:
I'd personally agree, but haven't seen any suggestions that encompass that ideology properly. 

As I said, if you have any suggestions, I'm all ears. 🙂

Reply:
Personally, I think a more "powershell-y" solution would look a bit like this:

```powershell
$Var -invoke Method($args)
```

But then we have a stark deviance from the current precedence we're setting with ternary and now these proposed `??` and `??=` operators.

Consider also that our _only existing_ combination assignment operators are of the form `+=`, `-=`, and similar (do we support others? I don't use many others personally). So then if we have, for example, something like this:

```powershell
$value = $null
$value -else 10 # returns 10
```

Then by extension we probably should implement a mirror assignment operator, which off the cuff may look like this: 

```powershell
$value -else:= 10
```

Or some variant thereof (opening up a general syntax for which would be more than welcome, allowing us to have more assignment operators, like `-replace:=` or some such).

This isn't bad, but I don't feel like it's optimal either. But it _is_ much more expressive. And if we want to go that route I'd agree that it's a good way to go. But so far it seems as though we're starting from what C# programmers will find familiar... and from _that_ context alone, I can say we've reached a point where I think the most sensible way forward is to make the minor breaking change and remove `?` from the standard set of variable name characters.

Reply:
Forcing the use of the more complicated variable syntax `${name}` is going to leave this feature underutilized. For example, I just asked another person on my team with 7 years of PowerShell experience if he knew about `${name}` syntax and he didn't. The current design will cause issues with scripts containing both `$name` and `${name}` when finding, renaming, or using variables.

I think we need to make a compromise, variables containing or ending with a `?` and implementing this operator with C# syntax. To support the most wide spread adoption of this operator without making people learn and support a more complicated syntax, variable names containing or ending with `?` should be removed when using the standard syntax `$name`, but still supported using the complicated syntax `${name}`. Implementing the operator this way gives people who want to use `?` a way to continue to do so while also giving the vast majority of users the same syntax with the operator.

Reply:
Lots of operators: -match, -replace, -contains, -split, -join, -as, -is, -isnot, -f, +=, -=, *+, /=, ++, --, etc. etc. The arithmetic operators very C-family based; the others - not so much. Some supporting very _odd_ syntax already (I'm looking at replace and join as I write that).

If someone is willing to looking at something more verbose, then something like:

`$varThatMayBeNull -isNull $FallBackValue`

seems fairly obvious. Similarly,

`$varThatMayHaveNullProperty -isPropertyNull $PropertyName, $FallBackValue`

also seems fairly obvious (with something shorter -ispropnull also reasonably verbose).

I currently handle the first with an infix if statement, e.g.:

`if( $varThatMayBeNull ) { $varThatMayBeNull } else { $FallBackValue } `

The second is more complex, which I would love to replace:

```
function SafeGetProp
{
    Param
    (
        [Object] $obj,
        [String] $propName
    )

    ## $prop = if( Get-Member -Name $propName -InputObject $obj -MemberType Property ) { $obj.$propName } else { $null }
    ## faster:
    $prop = &{ Set-StrictMode -Off; $obj.$propName }
    $prop
}

$prop = SafeGetProp $varThatMayHaveNullProperty $propertyName
if( $null -eq $prop ) { $prop = $FallBackValue }
```

Reply:
`-isNull` sounds like it's expected to return a boolean value; I'd probably avoid that.

Reply:
`-isValueNull`. It's flexible. :-) I'm throwing something out there that feels more like PowerShell to me. And is actually readable by average humans.

Reply:
`-is<anything>` sounds like it's expected to give a boolean result. For example `-isNull $null` I would expect to get `$true`, and I don't think that'd be a particularly uncommon pitfall of logic. 🙂 

But I think that is more in the right direction, yeah. I rather like `$value -else $fallback` just because it _kind of_ has the right feel to it (still far from perfect, but I don't think that we'll find a 'perfect' solution, really), and also because it can tie in with `$bool -then $trueValue -else $falseValue` -- though it's perfectly possible that may not be desirable.

Reply:
@SteveL-MSFT To respond to your question asking if anyone has other syntax suggestions that wouldn't result in breaking changes, I was tinkering a little today, and wondered what if we look at `!` as a postfix operator that is similar to the dot-reference operator, but only evaluating the right-hand side if the left-hand side is not null?

As far as I can tell, these are non-breaking changes:

```powershell
# This references the Length member, but only if $x -ne $null
# (no `.` required, since there's already one at the bottom of the `!`, but
# we could use `!.` here if we really felt it was necessary)
$x!Length

# This references the item at index 2, but only if $x -ne $null
$x![2]
```

That would be non-breaking. I also like the visual relationship between the dot-reference operator (`.`) and the non-null dot-reference operator (`!`). The actual operators here would be `!` and `![]`.

When comparing this with the other C# operators (ternary, null-coalescing, etc.), there's no conflict in thought and they are all explainable:

* `?:` is used for shorthand conditionals (logical expressions)
* `??` and `??=` both show that `??` is used to invoke an expression if a value is null
* `!` is used to invoke a member or to retrieve a collection item if a value is not null.